src/main/java/com/google/edith/servlets/DealsServlet.java

```diff
@@ -0,0 +1,51 @@
+package com.google.edith.servlets;
+
+import com.google.edith.DealItem;
+import com.google.edith.GroceryDataReader;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.io.BufferedReader;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that takes the user input from the receipt
+  * form and uses it to retrieve the best deal. */
+@WebServlet("/receipt-deals")
+public class DealsServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    StringBuilder sb = new StringBuilder();
+    BufferedReader reader = request.getReader();
+    try {
```

Comment:
I think you could use a try with resources block here, which will take care of closing any resource for you (in this case the BufferedReader) and should allow you to remove the finally block.

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html